tests/datasets/test_transfer.py

```diff
@@ -0,0 +1,229 @@
+TEST_AWS_UPLOAD_ID = 123
+
+TEST_ETAG = "5d24e152bcdfa5a0357f46471be3be6c"
+TEST_COMPLETED_PART = CompletedUploadPart(1, TEST_ETAG)
+
+OTHER_ETAG = "d084dd881a190aa5ffdf0ce21cff9509"
+OTHER_COMPLETED_PART = CompletedUploadPart(2, OTHER_ETAG)
+
+TEST_CONTENT = "".join(
+    random.choice(string.printable) for _ in range(2000)
+).encode("utf8")
+
+
+@pytest.fixture
+def mock_client_download(mocker, requests_mock):
+    object_client = mocker.Mock()
+    object_client.presign_download.return_value = TEST_URL
+
+    requests_mock.get(TEST_URL, content=TEST_CONTENT)
+
+    yield object_client
+
+    object_client.presign_download.assert_called_once_with(
+        PROJECT_ID, TEST_PATH
+    )
+
+
+@pytest.fixture
+def mock_client_upload_s3(mocker, requests_mock):
+    presigned_response_mock = mocker.Mock()
+    presigned_response_mock.provider = CloudStorageProvider.S3
+    presigned_response_mock.upload_id = TEST_AWS_UPLOAD_ID
+
+    object_client = mocker.Mock()
+    object_client.presign_upload.return_value = presigned_response_mock
+
+    yield object_client
+
+
+@pytest.fixture
+def mock_client_upload_gcs(mocker, requests_mock):
+    presigned_response_mock = mocker.Mock()
+    presigned_response_mock.provider = CloudStorageProvider.GCS
+    presigned_response_mock.url = TEST_URL
+
+    object_client = mocker.Mock()
+    object_client.presign_upload.return_value = presigned_response_mock
+
+    yield object_client
+
+
+def _assert_contains(dict_1, dict_2):
+    for key, value in dict_2.items():
+        assert dict_1[key] == value
+
+
+def test_download(mock_client_download):
+    assert (
+        transfer.download(mock_client_download, PROJECT_ID, TEST_PATH)
+        == TEST_CONTENT
+    )
+
+
+def test_download_stream(mock_client_download):
+    stream = transfer.download_stream(
+        mock_client_download, PROJECT_ID, TEST_PATH
+    )
+    assert b"".join(stream) == TEST_CONTENT
+
+
+def test_download_file(mock_client_download, tmpdir):
+    destination = tmpdir.join("destination.txt")
+
+    transfer.download_file(
+        mock_client_download, PROJECT_ID, TEST_PATH, destination
+    )
+    assert destination.read(mode="rb") == TEST_CONTENT
+
+
+def test_chunking_and_labelling_data_of_exact_sizes(mocker):
+    mocker.patch("faculty.datasets.transfer.UPLOAD_CHUNK_SIZE", 4)
+    content = [b"1111", b"2222", b"last"]
+    a = transfer._rechunk_and_label_as_last(content)
+    assert list(a) == [(b"1111", False), (b"2222", False), (b"last", True)]
+
+
+def test_chunking_and_labelling_of_smaller_sizes(mocker):
+    mocker.patch("faculty.datasets.transfer.UPLOAD_CHUNK_SIZE", 12)
+    content = [b"1111", b"2222", b"last"]
+    a = transfer._rechunk_and_label_as_last(content)
+    assert list(a) == [(b"11112222last", True)]
+
+
+def test_chunking_and_labelling_of_greater_sizes(mocker):
+    mocker.patch("faculty.datasets.transfer.UPLOAD_CHUNK_SIZE", 4)
+    content = [b"11112222last"]
+    a = transfer._rechunk_and_label_as_last(content)
```

Comment:
As above:
"a" isn't a great variable name. Can we change it to something like "chunks"?